execution-engine/contracts/hdac-system/pop/src/contract_votes.rs

```diff
@@ -0,0 +1,225 @@
+        let mut votes = BTreeMap::new();
+        for (name, _) in runtime::list_named_keys() {
+            let mut split_name = name.split('_');
+            if Some("a") != split_name.next() {
+                continue;
+            }
+
+            let to_publickey = |hex_str: &str| -> Result<PublicKey> {
+                if hex_str.len() != 64 {
+                    return Err(Error::VoteKeyDeserializationFailed);
+                }
+                let mut key_bytes = [0u8; 32];
+                let _bytes_written = base16::decode_slice(hex_str, &mut key_bytes)
+                    .map_err(|_| Error::VoteKeyDeserializationFailed)?;
+                debug_assert!(_bytes_written == key_bytes.len());
+                Ok(PublicKey::from(key_bytes))
+            };
+
+            let to_hash = |hex_str: &str| -> Result<Key> {
+                if hex_str.len() != 64 {
+                    return Err(Error::VoteKeyDeserializationFailed);
+                }
+                let mut hash_bytes = [0u8; 32];
+                let _bytes_written = base16::decode_slice(hex_str, &mut hash_bytes)
+                    .map_err(|_| Error::VoteKeyDeserializationFailed)?;
+                debug_assert!(_bytes_written == hash_bytes.len());
+                // TODO: How to convert hash to key
+                Ok(Key::Hash(hash_bytes))
+            };
+
+            let hex_key = split_name
+                .next()
+                .ok_or(Error::VoteKeyDeserializationFailed)?;
+            let dapp_user = to_publickey(hex_key)?;
+
+            let hex_key = split_name
+                .next()
+                .ok_or(Error::VoteKeyDeserializationFailed)?;
+            let dapp_owner = to_hash(hex_key)?;
+
+            let balance = split_name
+                .next()
+                .and_then(|b| U512::from_dec_str(b).ok())
+                .ok_or(Error::VotesDeserializationFailed)?;
+
+            votes.insert(
+                VoteKey {
+                    user: dapp_user,
+                    dapp: dapp_owner,
+                },
+                balance,
+            );
+        }
+
+        Ok(Votes(votes))
+    }
+
+    /// Writes the current stakes to the contract's known urefs.
+    pub fn write(votes: &Votes) {
+        // Encode the stakes as a set of uref names.
+        let mut new_urefs: BTreeSet<String> = votes
+            .0
+            .iter()
+            .map(|(vote_key, balance)| {
+                let to_hex_string = |address: PublicKey| -> String {
+                    let bytes = address.value();
+                    let mut ret = String::with_capacity(64);
+                    for byte in &bytes[..32] {
+                        write!(ret, "{:02x}", byte).expect("Writing to a string cannot fail");
+                    }
+                    ret
+                };
+
+                let to_hex_string_from_hash = |hash: Key| -> String {
+                    let bytes = hash.to_bytes().unwrap_or_revert_with(ApiError::Deserialize);
```

Comment:
If `Hash` is`to_bytes()`as shown below, the prefix are appended to the value of 32 bytes. Is this what you intended?

If it is intended, the data of this value will be 33 bytes, not 32 bytes. Is it okay to write only 32 bytes as a string?
```
// example
hash : [1, 2, 3, 4, 5, 6, 7, ...., 32] -> length : 32
hash.to_bytes() : [1, 1, 2, 3, 4, 5, 6, 7, ... 32] -> length : 33
```

Reply:
Ehh?? I'm not intended about that, and I now know that thanks to your checking.
@jhyeom26 says that ContractRef should be stored as hash, not URef, and find for how to store and Key would be appropriate..

Reply:
The first byte would be indicator of the type of the data (hash, uref, ..)

Reply:
Yes, the first byte is a prefix indicating address, hash, uref, and local. If it is used, read and write should be modified to 33 bytes.
In addition, the currently value is [1,1,2,3,4,5,6,7, ... 31] -> length : 32.

Reply:
```
let bytes = hash.into_hash().unwrap_or_revert_with(ApiError::Deserialize);
```
Would it be good to change that?

Reply:
Definition of `into_hash()`
```rust
/// Returns the inner bytes of `self` if `self` is of type [`Key::Hash`], otherwise returns
    /// `None`.
    pub fn into_hash(self) -> Option<[u8; KEY_HASH_LENGTH]> {
        match self {
            Key::Hash(hash) => Some(hash),
            _ => None,
        }
    }
```

Reply:
I am not sure exactly, but can you use the `from_bytes` in `key.rs`?

If we can't use it, the example above looks good too.

Reply:
As we talked with slack, we go forward with `into_hash ()`